website/docs/tstruct.md

```diff
@@ -124,11 +191,286 @@ class ChildTwo < T::Struct
+  include Common
+  prop :foo, Integer
+  prop :quz, Symbol
+end
+```
+
 If the code absolutely must use inheritance and cannot use composition, either:
 
 - Avoid using `T::Struct`, and instead define a normal class, with things like
   `attr_reader` and an explicit `initialize` method.
 
-- Change the superclass from `T::Struct` to `T::InexactStruct`. Sorbet will no
-  longer check the types of any arguments passed to the `initialize` method on
-  the subclass. This should only be used as a last resort.
+- Change the superclass from `T::Struct` to `T::InexactStruct`. This causes
+  Sorbet to no longer statically check the types of any arguments passed to the
+  `initialize` method on the subclass, but does allow defining `T::Struct`
+  hierarchies. This should only be used as a last resort.
+
+<br>
+
+---
+
+<br>
+
+## Legacy code and historical context
+
+The `prop` DSL used by `T::Struct` predates Sorbet by about 5 years. It was
+originally conceived at Stripe in early 2013 to form the basis for Stripe's
+internal [object-document mapper][odm] (ODM). By the time Stripe began internal
+development on Sorbet in late 2017, Stripe's ODM was by far the most commonly
+used internal abstraction for associating types with methods. At a time when it
+was not clear that the as-yet-unnamed Ruby type checker project would succeed or
+not, we were eager to build on existing abstractions to bootstrap early type
+coverage.
+
+[odm]:
+  https://en.wikipedia.org/wiki/Object%E2%80%93relational_mapping#Object-oriented_databases
+
+A decision was made to factor the code for the `prop` DSL into a standalone
+library, to allow using it independently of the database-specific code in
+Stripe's ODM library. From this effort, `T::Struct` was born. A `T::Struct` is
+essentially a Stripe [database model] class without the database.
+
+[database model]: https://en.wikipedia.org/wiki/Database_model
+
+Unfortunately, this process left warts in the publicly-accessible `T::Struct`
+APIs that persist today. Certain parts of the `prop` DSL only make sense when
+used alongside Stripe-internal abstractions. The DSL also contains things that
+are technically publicly accessible that were never meant to be. This legacy
+makes it hard to evolve and improve the `T::Struct` APIs without breaking
+existing code.
+
+The remainder of this documentation is presented for completeness. Use the APIs
+below at your own discretion. Our goal here is simply to outline the potential
+pitfalls that arise when using them.
+
+## `serialize` and `from_hash`: Converting `T::Struct` to and from `Hash`
+
+It's possible to convert a `T::Struct` instance to and from `Hash` instances:
+
+```ruby
+class A < T::Struct
+  prop :foo, Integer
+end
+
+# (1) `serialize` converts from `T::Struct` to `Hash`
+serialized = A.new(foo: 42).serialize
+p(serialized) # => {"foo"=>42}
+
+# (1) `from_hash` converts from `Hash` to `T::Struct`
+deserialized = A.from_hash(serialized)
+p(deserialized) # => <A foo=42>
```

Comment:
Should this section mention `from_hash!`?

Reply:
Sure, I can mention it.

Note to self: `from_hash!` is `from_hash(..., /* strict */true)` which requires that there be no unexpected keys in the Hash, but still doesn't do runtime type checking.